pandas/core/ops.py

```diff
@@ -257,30 +256,87 @@ def add_flex_arithmetic_methods(cls, flex_arith_method,
                 exclude=exclude)
 
 
-class _TimeOp(object):
+class _Op(object):
+
     """
-    Wrapper around Series datetime/time/timedelta arithmetic operations.
-    Generally, you should use classmethod ``maybe_convert_for_time_op`` as an
-    entry point.
+    Wrapper around Series arithmetic operations.
+    Generally, you should use classmethod ``_Op.get_op`` as an entry point.
+
+    This validates and coerces lhs and rhs depending on its dtype and
+    based on op. See _TimeOp also.
+
+    Parameters
+    ----------
+    left : Series
+        lhs of op
+    right : object
+        rhs of op
+    name : str
+        name of op
+    na_op : callable
+        a function which wraps op
     """
-    fill_value = iNaT
+
+    fill_value = np.nan
     wrap_results = staticmethod(lambda x: x)
     dtype = None
 
     def __init__(self, left, right, name, na_op):
+        self.left = left
+        self.right = right
+
+        self.name = name
+        self.na_op = na_op
+
+        self.lvalues = left
+        self.rvalues = right
+
+    @classmethod
+    def get_op(cls, left, right, name, na_op):
+        """
+        Get op dispatcher, returns _Op or _TimeOp.
+
+        If ``left`` and ``right`` are appropriate for datetime arithmetic with
+        operation ``name``, processes them and returns a ``_TimeOp`` object
+        that stores all the required values.  Otherwise, it will generate
+        either a ``_Op``, indicating that the operation is performed via
+        normal numpy path.
+        """
+        is_timedelta_lhs = is_timedelta64_dtype(left)
+        is_datetime_lhs = (is_datetime64_dtype(left) or
+                           is_datetime64tz_dtype(left))
 
-        # need to make sure that we are aligning the data
         if isinstance(left, ABCSeries) and isinstance(right, ABCSeries):
-            left, right = left.align(right, copy=False)
+            # avoid repaated alignment
+            if not left.index.equals(right.index):
+                left, right = left.align(right, copy=False)
+
+                index, lidx, ridx = left.index.join(right.index, how='outer',
+                                                    return_indexers=True)
+                # convert DatetimeIndex to have UTC
```

Comment:
Updated the comment to explain below:

If the ``Series/DataFrame`` have different tz, the indexes are converted to UTC (on 0.18.1). Added tests also.

```
base = pd.DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03'], tz='UTC')
idx1 = base.tz_convert('Asia/Tokyo')[:2]
idx2 = base.tz_convert('US/Eastern')[1:]

pd.Series([1, 2], index=idx1) + pd.Series([1, 1], index=idx2)
# 2011-01-01 00:00:00+00:00    NaN
# 2011-01-02 00:00:00+00:00    3.0
# 2011-01-03 00:00:00+00:00    NaN
# dtype: float64
```